jax/_src/scipy/stats/_core.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import namedtuple
+from functools import partial
+from typing import Optional, List, Tuple, Iterable, Any
+
+import scipy
+
+from jax import jit, vmap
+import jax.numpy as jnp
+from jax._src.numpy.util import _wraps
+from jax._src.numpy.lax_numpy import _check_arraylike
+
+ModeResult = namedtuple("ModeResult", ("mode", "count"))
+
+
+@partial(jit, static_argnums=(1,))
+def _mode(x: Iterable[Any], axis: Optional[int]) -> ModeResult:
+  """
+  Hidden jit-compliable function accessed through `jax.scipy.stats.mode`.
+  """
+  def _mode_helper(x: Iterable[Any]) -> Tuple[Iterable[Any], Iterable[Any]]:
+    vals, counts = jnp.unique(x, return_counts=True, size=x.size)
+    return vals[jnp.argmax(counts)], jnp.max(counts)
+
+  if x.size == 0:
+    return ModeResult(jnp.array([]), jnp.array([]))
+  elif axis is None:
+    vals, counts = _mode_helper(x)
+    return ModeResult(vals.reshape(-1), counts.reshape(-1))
+  else:
+    new_shape: List[int] = list(x.shape)
+    new_shape[axis] = 1
+    x = jnp.moveaxis(x, axis, 0).reshape(x.shape[axis], -1)
+    vals, counts = vmap(_mode_helper, in_axes=(1,))(x)
+    return ModeResult(vals.reshape(new_shape), counts.reshape(new_shape))
+
+
+@_wraps(scipy.stats.mode)
+def mode(x: Iterable[Any], axis: int = 0, nan_policy: str = "propagate") -> ModeResult:
+  _check_arraylike("mode", x)
```

Comment:
After `_check_arraylike`, please call `x = jnp.asarray(x)` to ensure that `x` is converted to a `DeviceArray`.

Also, there is a `canonicalize_axis` utility used elsewhere that I think would be relevant here. Regarding handling of `axis=None`, what would you think about doing it this way, so that `axis` can always be an integer in the helper? https://github.com/google/jax/blob/f0b139a59a749f6b3a4776acc50c7a2fa57e9a9e/jax/_src/numpy/reductions.py#L565-L567

A benefit of that I think is that you'd end up calling the `_mode_helper` code in only a single location, so it may lead to a cleaner implementation.

Reply:
I'll trust your thoughts here. My only concern would be the performance implications.

An exaggerated example being calling `mode` successively in a loop, always with `axis=None` with a large `x`.

Unfortunately, I have a Radeon VII on this machine, so I don't think I could profile that appropriately right now.